lib/puppet/rest_client/client.rb

```diff
@@ -0,0 +1,76 @@
+# frozen_string_literal: true
+require 'httpclient'
+require 'puppet'
+require 'puppet/file_system'
+
+module Puppet
+  module Rest
+    # TBD
+    class Client
+      def self.instance
+        @@instance ||= new(Puppet.settings[:http_user_agent],
+                           Puppet::PUPPETVERSION)
+      end
+
+      # Create an HTTP client that makes requests with a default header
+      # @param [String] user_agent the User-Agent string for the default header
+      # @param [String] version the version of puppet in use, for the default header
+      def initialize(user_agent, version)
+        @http = HTTPClient.new(
+          agent_name: nil,
+          default_header: {
+            'User-Agent' => user_agent,
+            'X-PUPPET-VERSION' => version
+          }
+
+        )
+
+        # enable to see traffic on the wire
+        #@http.debug_dev = $stderr
+
+        @http.tcp_keepalive = true
+        @http.connect_timeout = 10
+        @http.receive_timeout = 60 * 60
+        @http.request_filter << self
+      end
+
+      def disconnect
+        @http.reset_all
+      end
+
+      def filter_request(req)
+        Puppet.debug _("Connecting to %{uri} (%{method})") % {uri: req.header.request_uri, method: req.header.request_method }
+      end
+
+      def filter_response(_req, res)
+        Puppet.debug _("Done %{status} %{reason}\n\n") % { status: res.status, reason: res.reason }
+      end
+
+      def check_validation_mode
+        ssl_config = @http.ssl_config
+
+        # Is this expensive to do for every request? Probably not compared to the
+        # shenanigans we're doing now
+        if Puppet::FileSystem.exist?(Puppet.settings[:hostcert])
+          ssl_config.clear_cert_store
+          ssl_config.verify_mode = OpenSSL::SSL::VERIFY_PEER
+          ssl_config.add_trust_ca(Puppet::FileSystem.expand_path(Puppet.settings[:localcacert]))
+          ssl_config.set_client_cert_file(Puppet::FileSystem.expand_path(Puppet.settings[:hostcert]),
+                                          Puppet::FileSystem.expand_path(Puppet.settings[:hostprivkey]))
+          ssl_config.add_crl(Puppet::FileSystem.expand_path(Puppet.settings[:hostcrl]))
+        else
+          ssl_config.clear_cert_store
+          ssl_config.verify_mode = OpenSSL::SSL::VERIFY_NONE
+        end
+      end
```

Comment:
I think it can be expensive to reparse/rebuild the cert store for each request, especially since the state won't be changing during normal agent operation, only during the agent ssl bootstrap phase.

Can we instead make the client always `VERIFY_PEER` in its `initialize` and provide a different client instance for unsafe connections?

I think we're going to have a similar problem when we want to trust the puppet CA and the default system store for https based file sources, report servers, etc, in which case we might need three different clients?

Reusing the same client across different security domains seems like we're asking for trouble, especially if wanted to do multi-threaded pluginsync.

Reply:
Our current validator uses the `ssl_store` from `Puppet::SSL::Host` (which is cached for the lifetime of the ssl host), is it possible to continue using the host's ssl_store when creating connections and/or allowing us to pass a store in from elsewhere?

Also, we don't have to do that as part of this work, necessarily but I think this work should allow us to do so relatively easily in the next ticket when have to download the CRL. See https://github.com/puppetlabs/puppet/pull/6797#discussion_r184236925 and https://github.com/puppetlabs/puppet/pull/6797/files#diff-1188cf806577ab933977ff429b3bc1b9R708